src/metabase/lib/card.cljc

```diff
@@ -72,45 +74,55 @@
-
-  ([metadata-providerable :- ::lib.schema.metadata/metadata-providerable
-    card-or-id            :- [:maybe [:or ::lib.schema.id/card ::lib.schema.metadata/card]]
-    col                   :- :map]
+(defn- ->card-metadata-column
+  "Massage possibly-legacy Card results metadata into MLv2 ColumnMetadata. Note that `card` might be unavailable so we
+  accept both `card-id` and `card`."
+  [col
+   card-id
+   card
+   field]
    (let [col (-> col
                  (update-keys u/->kebab-case-en))]
      (cond-> (merge
               {:base-type :type/*, :lib/type :metadata/column}
-              (when-let [field-id (:id col)]
-                (try
-                  (lib.metadata/field metadata-providerable field-id)
-                  (catch #?(:clj Throwable :cljs :default) _
-                    nil)))
+              field
               col
               {:lib/type                :metadata/column
                :lib/source              :source/card
                :lib/source-column-alias ((some-fn :lib/source-column-alias :name) col)})
-       card-or-id
-       (assoc :lib/card-id (u/the-id card-or-id))
+       card-id
+       (assoc :lib/card-id card-id)
 
        (and *force-broken-card-refs*
             ;; never force broken refs for Models, because Models can have give columns with completely
             ;; different names the Field ID of a different column, somehow. See #22715
             (or
-             ;; we can only do this check if `card-or-id` is passed in.
-             (not card-or-id)
-             (not= (:type (lib.metadata/card metadata-providerable (u/the-id card-or-id)))
-                   :model)))
+             ;; we can only do this check if `card-id` is passed in.
+             (not card-id)
+             (not= (:type card) :model)))
        (assoc ::force-broken-id-refs true)
 
        ;; If the incoming col doesn't have `:semantic-type :type/FK`, drop `:fk-target-field-id`.
        ;; This comes up with metadata on SQL cards, which might be linked to their original DB field but should not be
        ;; treated as FKs unless the metadata is configured accordingly.
        (not= (:semantic-type col) :type/FK)
-       (assoc :fk-target-field-id nil)))))
+       (assoc :fk-target-field-id nil))))
+
+(mu/defn ->card-metadata-columns :- [:sequential ::lib.schema.metadata/column]
+  "Massage possibly-legacy Card results metadata into MLv2 ColumnMetadata."
+  ([metadata-providerable cols]
+   (->card-metadata-columns metadata-providerable nil cols))
+
+  ([metadata-providerable :- ::lib.schema.metadata/metadata-providerable
+    card-or-id            :- [:maybe [:or ::lib.schema.id/card ::lib.schema.metadata/card]]
+    cols                  :- [:sequential :map]]
+   (let [metadata-provider (lib.metadata/->metadata-provider metadata-providerable)
+         card-id           (u/the-id card-or-id)
+         card              (lib.metadata/card metadata-providerable card-id)
```

Comment:
It doesn't seem like a good idea calling this with `nil`.

Reply:
Fixed